scylla-cdc-driver3/src/main/java/com/scylladb/cdc/cql/driver3/Driver3WorkerCQL.java

```diff
@@ -147,28 +197,99 @@ public void onFailure(Throwable t) {
                 // returned by the preparing of the query)
                 // and fetching of the base table schema (now)
                 // there was at most a single schema change.
-                Row row = rs.one();
+                Row row = partialStreamResult.one();
                 if (schema == null) {
                     try {
                         schema = Driver3SchemaFactory.getChangeSchema(row, session.getCluster().getMetadata());
                     } catch (Driver3SchemaFactory.UnresolvableSchemaInconsistencyException ex) {
-                        fut.completeExceptionally(ex);
+                        result.completeExceptionally(ex);
                         return;
                     }
                 }
                 Driver3RawChange newChange = new Driver3RawChange(row, schema);
+                result.complete(Optional.of(newChange));
+            }
+        }
 
-                // lastChangeId determines the point from which we should
-                // start reading within a window. In this implementation
-                // we simply read the entire window and skip rows that
-                // were before lastChangeId.
-                //
-                // If lastChangeId is Optional.empty(), then we read
-                // the entire window.
-                if (!lastChangeId.isPresent() || newChange.getId().compareTo(lastChangeId.get()) > 0) {
-                    fut.complete(Optional.of(newChange));
+        private void startReadingEntireStreamsResult(CompletableFuture<Optional<RawChange>> result) {
+            Stream<StreamId> streams = task.streams.stream();
+
+            if (task.state.getLastConsumedChangeId().isPresent()) {
+                ChangeId lastConsumedChangeId = task.state.getLastConsumedChangeId().get();
+                streams = streams.filter(s -> s.compareTo(lastConsumedChangeId.getStreamId()) > 0);
```

Comment:
Is compareTo on StreamId doing the right thing? Is it using the same sorting order as a select would use? That means the order of tokens of partition keys.

Reply:
Yes, as far as I tested this did the right thing. Also the integration tests verify that assumption.